File: backend/src/main/java/com/ecms/service/FeedbackService.java
package com.ecms.service;

import com.ecms.dto.request.FeedbackRequest;
import com.ecms.dto.response.FeedbackResponse;

import java.util.List;
import java.util.Map;

/**
 * @author  ThangNB - HE201024
 * @created 2026-07-19
 * @updated 2026-07-20
 *
 * Patient feedback contract for UC-48 (Submit Feedback); the stored records
 * feed the Clinic Manager's report in UC-53.
 *
 * Business rules: BR-21 — one feedback per appointment.
 */
public interface FeedbackService {

    /**
     * Records a patient's feedback for one of their own visits
     * (UC-48 normal flow steps 2-5).
     *
     * @param patientId the authenticated patient, resolved from their account
     * @param request   rating, optional comment and per-participant scores
     * @return the stored feedback with status PENDING
     *
     * Validate: UC-48 PRE-2 (the appointment must be COMPLETED), PRE-3 /
     * BR-21 (no feedback may already exist for it) and ownership — a patient
     * can only rate their own visit.
     */
    FeedbackResponse submitFeedback(Long patientId, FeedbackRequest request);

    /**
     * Lists the feedback a patient has already submitted.
     *
     * @param patientId the authenticated patient
     * @return their own feedback only
     */
    List<FeedbackResponse> getMyFeedbacks(Long patientId);

    /**
     * Describes a visit and everyone who took part in it — doctor,
     * receptionist, lab technician — so the feedback form can offer a rating
     * per person.
     *
     * @param patientId     the authenticated patient
     * @param appointmentId the visit being rated
     * @return visit summary plus the participant list
     *
     * Validate: ownership — the appointment must belong to {@code patientId},
     * otherwise a patient could enumerate other people's visits.
     */
    Map<String, Object> getVisitParticipants(Long patientId, Long appointmentId);
}

File: frontend/src/services/feedbackService.js
/**
 * @author  ThangNB - HE201024
 * @created 2026-07-19
 * @updated 2026-07-20
 *
 * API client for patient feedback (UC-48 Submit Feedback).
 * Every endpoint is scoped to the signed-in patient server-side, so no
 * patient id is ever sent from here.
 */
import axiosClient from '../api/axiosClient'

export const feedbackService = {
  /**
   * Submits feedback for a completed visit.
   * @param {{appointmentId:number, rating:number, content?:string, isAnonymous?:boolean}} data
   * @returns {Promise} the stored feedback (status PENDING)
   *
   * Validate: rating is required and must be 1..5, checked in the form and
   * again by the backend (UC-48 E1). BR-21 (one feedback per appointment) is
   * enforced server-side and surfaces here as an error response.
   */
  submit: (data) => axiosClient.post('/v1/feedbacks', data),

  /**
   * Lists the signed-in patient's own submitted feedback.
   * @returns {Promise} that patient's feedback only
   */
  getMy: () => axiosClient.get('/v1/feedbacks/my'),

  /**
   * Loads the staff who took part in a visit (doctor, receptionist, lab
   * technician) so the form can offer a rating per person.
   * @param {number} appointmentId the visit being rated
   * @returns {Promise} visit summary plus participant list
   */
  getParticipants: (appointmentId) =>
    axiosClient.get(`/v1/feedbacks/appointment/${appointmentId}/participants`),
}
